File: android/src/main/java/y2k/joyreactor/platform/AndroidNavigation.java
package y2k.joyreactor.platform;

import android.app.Activity;
import android.app.Application;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import y2k.joyreactor.*;

/**
 * Created by y2k on 10/19/15.
 */
public class AndroidNavigation extends Navigation {

    static Post sPostArgument; // FIXME:
    static String sPostIdArgument = "2294127"; // FIXME:
    Activity currentActivity;

    public AndroidNavigation(Application app) {
        app.registerActivityLifecycleCallbacks(new MyActivityLifecycleCallbacks());
    }

    @Override
    public void switchProfileToLogin() {
        currentActivity.startActivity(new Intent(currentActivity, LoginActivity.class));
        currentActivity.finish();
    }

    @Override
    public void switchLoginToProfile() {
        currentActivity.startActivity(new Intent(currentActivity, ProfileActivity.class));
        currentActivity.finish();
    }

    @Override
    public void closeCreateComment() {
        // TODO:
    }

    @Override
    public void closeAddTag() {
        AddTagDialogFragment.dismiss((AppCompatActivity) currentActivity);
    }

    @Override
    public void openPost(Post post) {
        sPostArgument = post;
        currentActivity.startActivity(new Intent(currentActivity, PostActivity.class));
    }

    @Override
    public Post getArgumentPost() {
        return sPostArgument;
    }

    @Override
    public void openPost(String postId) {
        sPostIdArgument = postId;
        currentActivity.startActivity(new Intent(currentActivity, PostActivity.class));
    }

    @Override
    public String getArgumentPostId() {
        return sPostIdArgument;
    }

    @Override
    public void openBrowser(String url) {
        currentActivity.startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(url)));
    }

    @Override
    public void openVideo(String postId) {
        // TODO:
        sPostIdArgument = postId;
        currentActivity.startActivity(new Intent(currentActivity, VideoActivity.class));
    }

    @Override
    public void openImageView(Post post) {
        // TODO:
    }

    private class MyActivityLifecycleCallbacks implements Application.ActivityLifecycleCallbacks {

        @Override
        public void onActivityResumed(Activity activity) {
            currentActivity = activity;
        }

        @Override
        public void onActivityPaused(Activity activity) {
            currentActivity = null;
        }

        @Override
        public void onActivityCreated(Activity activity, Bundle savedInstanceState) {
        }

        @Override
        public void onActivityStarted(Activity activity) {
        }

        @Override
        public void onActivityStopped(Activity activity) {
        }

        @Override
        public void onActivitySaveInstanceState(Activity activity, Bundle outState) {
        }

        @Override
        public void onActivityDestroyed(Activity activity) {
        }
    }
}

File: android/src/main/kotlin/y2k/joyreactor/platform/AndroidNavigation.kt
package y2k.joyreactor.platform

import android.app.Activity
import android.app.Application
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.support.v4.app.FragmentManager
import android.support.v7.app.AppCompatActivity
import y2k.joyreactor.CreateCommentFragment
import y2k.joyreactor.PostLikeFragment
import y2k.joyreactor.common.ActivityLifecycleCallbacksAdapter
import y2k.joyreactor.common.platform.NavigationService
import y2k.joyreactor.common.startActivity
import y2k.joyreactor.viewmodel.CreateCommentViewModel
import y2k.joyreactor.viewmodel.PostLikeViewModel
import kotlin.reflect.KClass

/**
 * Created by y2k on 10/19/15.
 */
class AndroidNavigation(app: Application) : NavigationService {

    private val fragmentManager: FragmentManager?
        get() = (currentActivity as AppCompatActivity?)?.supportFragmentManager

    internal var currentActivity: Activity? = null

    init {
        app.registerActivityLifecycleCallbacks(MyActivityLifecycleCallbacks())
    }

    override fun close() {
        val dialog = fragmentManager?.findFragmentByTag("dialog")
        if (dialog == null) currentActivity?.finish()
        else (dialog as DialogFragment).dismiss()
    }

    override fun <T : Any> open(vmType: KClass<T>, argument: String) {
        sArgument = argument
        when (vmType) {
            PostLikeViewModel::class -> PostLikeFragment().show(fragmentManager, "dialog")
            CreateCommentViewModel::class -> CreateCommentFragment().show(fragmentManager, "dialog")
            else -> generalOpen(vmType)
        }
    }

    private fun <T : Any> generalOpen(vmType: KClass<T>) {
        try {
            val activityClsName = vmType.java.simpleName.replace("ViewModel", "Activity")
            val activityCls = javaClass.classLoader.loadClass(
                currentActivity!!.packageName + "." + activityClsName)
            startActivity(activityCls.kotlin)
        } catch (e: Exception) {
            throw Exception("Can't handler navigation to $vmType", e)
        }
    }

    private fun startActivity(activityType: KClass<*>) {
        currentActivity?.startActivity(activityType)
    }

    override val argument: String
        get() = sArgument

    override fun openBrowser(url: String) {
        currentActivity!!.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(url)))
    }

    private inner class MyActivityLifecycleCallbacks : ActivityLifecycleCallbacksAdapter() {

        override fun onActivityCreated(activity: Activity?, savedInstanceState: Bundle?) {
            currentActivity = activity
        }

        override fun onActivityResumed(activity: Activity?) {
            currentActivity = activity
        }

        override fun onActivityPaused(activity: Activity?) {
            if (currentActivity == activity) currentActivity = null
        }
    }

    companion object {

        // FIXME: перенести на уровень платформы
        internal var sArgument = "2294127"
    }
}